packages/imperative/src/rest/src/client/Proxy.ts

```diff
@@ -0,0 +1,178 @@
+     * @static
+     * @param session Zowe `ISession` containing the hostname for the http request.
+     * @returns `URL` to proxy server
+     * @memberof Proxy
+     */
+    public static getSystemProxyUrl(session: ISession): URL | undefined {
+        return this.getProxySettings(session)?.proxyUrl;
+    }
+
+    /**
+     * If the NO_PROXY or no_proxy environment variables are set with a comma separated
+     * list of hostnames it will try to match the hostname of the Zowe `ISession` and
+     * return `true` if found. Performs a simple string compare ignoring casing and white
+     * spaces, but will not resolve hostnames to ip addressees and not perform wildcard matching.
+     * @static
+     * @param session Zowe `ISession` containing the hostname for the http request.
+     * @returns `true` if the Zowe session host matches an entry in the comma separated
+     *          list of hostnames in the environment variable. `false` otherwise.
+     * @memberof Proxy
+     */
+    public static matchesNoProxySettings(session: ISession): boolean {
+        const noProxyValues = this.getNoProxyEnvVariables();
+        if (!noProxyValues) {
+            return false;
+        }
+        if (noProxyValues.includes(session.hostname.toLocaleLowerCase())) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Parses environment variables for proxy servers.
+     * @private
+     * @static
+     * @param session Zowe `ISession` containing the hostname for the http request.
+     * @returns instance of private `ProxySetting` or `undefined`
+     * @memberof Proxy
+     */
+    private static getProxySettings(session: ISession): ProxySetting | undefined {
+        if (this.matchesNoProxySettings(session)) {
+            return;
+        }
+        const protocol = session.protocol ?? HTTPS_PROTOCOL;
+        let envVariable: string | undefined;
+        if (protocol === HTTP_PROTOCOL) {
+            envVariable = this.getHttpEnvVariables();
+        }
+        else if (protocol === HTTPS_PROTOCOL) {
+            envVariable = this.getHttpsEnvVariables();
+        }
+        const proxyUrl = this.checkUrl(envVariable);
+        if (proxyUrl) {
+            return {proxyUrl, protocol};
+        }
+    }
+
+    /**
+     * Parses environment variables valid for http requests.
+     * @private
+     * @static
+     * @returns `string` if valid variable is found or undefined.
+     * @memberof Proxy
+     */
+    private static getHttpEnvVariables(): string | undefined {
+        return env.HTTP_PROXY ?? env.http_proxy ?? undefined;
+    }
+
+    /**
+     * Parses environment variables valid for https requests.
+     * @private
+     * @static
+     * @returns `string` if valid variable is found or undefined.
+     * @memberof Proxy
+     */
+    private static getHttpsEnvVariables(): string | undefined {
+        return env.HTTPS_PROXY ?? env.https_proxy ?? this.getHttpEnvVariables() ?? undefined;
```

Comment:
Since `this.getHttpEnvVariables` falls back to `undefined`, the `?? undefined` can be removed here as well